### ISSUE TITLE
Meta issue about ktsctl enhancements

### ISSUE THREAD

**Issue by hadronized** [hadronized/kak-tree-sitter#279]
I’ve been wanting to update ktsctl to support many new features. First thing first, let’s list the already existing feature requests:

- #34, #140: this is going to be easy, as I want to introduce a small data type to hold a language source (`Source` or something like that), that will be an `enum` with `GitRepo | LocalPath`, basically. Once we have that, we don’t have to do anything if it’s a local path, and if it’s a git repository, we just need to create a directory hierarchy in a cache folder and check whether it exists.
  - [x] That feature introduces a new problem: synchronization. Because a user can install some tree-sitter resources one day, and weeks / months later, would like to update. We will then need a flag to synchronize already installed grammars… Actually, it’s easy; we need to force git sources to have a pin, and then we can simply just compare.
  - [x] And a flag to check how much we are lagging behind would be nice too (comparing commits with the `pin`, for instance).
- #30: an old one, that I’ve been wanting for a long time (I sometimes check a new machine and I have to manually install everything again). Again, that shouldn’t be too hard to do as we already have the list of available languages in the config.
  - ~[ ] We can add a new option in the config to provide a language parallelism, that could default to something like `8`, to fetch, compile, link and install languages in parallel. Careful, as that can be tricky so special synchronization will be required when doing IO here.~ No, we won’t do that. Racing IO and git operations just sound like a super bad idea.

---

Now for new features I’m thinking of:

- [x] I think it could be nice to have subcommands `manage` and `info` to, respectively, manage resources (fetch, compile, link, install, uninstall, etc.) and get information about available languages.
- [x] Getting information could still include the already existing `--has`, but we could also add `--all` to list all available languages and provide information about them (are they installed? grammar? queries? display the associated configuration like `pin`, compilation flags, specific overrides like `remove_default_highlighter`, etc. etc.).

**Comment by vbauerster** [comment]
Please consider using `git clone --depth 1` (aka shallow clone) while fetching remote git repos with `ktsctl`. When connection is slow cloning heavy repos such as helix is likely to timeout, in fact it is happened to me with `ktsctl manage -f go` and there was no sign of error from `ktsctl` side.

**Comment by hadronized** [comment]
Good call.

**Comment by vbauerster** [comment]
Looks like `git clone --depth 1` can only fetch HEAD of master/main branch. If we still want to respect `pin` option then as per [this](https://stackoverflow.com/a/43136160) answer following git commands need to be run at once:

```
git init
git remote add origin <url>
git fetch --depth 1 origin <sha1>
git checkout FETCH_HEAD
```

**Comment by hadronized** [comment]
@vbauerster that should be fixed now.
